`ot_otp`: perform partition descrambling in DAI

luismarques opened the pull request:
The descrambling for the token loading path will be done in a separate PR.

luismarques committed:
[ot] hw/opentitan: ot_otp_eg: perform partition descrambling in DAI

Unscramble DAI reads from secret partitions, which are scrambled with a
scrambling key (configured per-partition as a netlist constant) via a
simple Present block cipher.

Co-authored-by: Alex Jones <alex.jones@lowrisc.org>
Signed-off-by: Luís Marques <luismarques@lowrisc.org>
Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

rivos-eblot commented on hw/opentitan/ot_otp_eg.c:
```diff
@@ -908,6 +908,12 @@ ot_otp_eg_lci_change_state_line(OtOTPEgState *s, OtOTPLCIState state, int line);
 #define ot_otp_hexdump(_s_, _b_, _l_)
 #endif
 
+OtOTPTokenValue scrambling_keys[] = {
```
All scrambling data are stored in the `earlgrey.cfg`  (or `darjeeling.cfg`) file, which is generated with the `cfggen.py`. Ping me on Slack if you need details/help.

rivos-eblot commented on hw/opentitan/ot_otp_eg.c:
```diff
@@ -1762,6 +1768,17 @@ static void ot_otp_eg_dai_read(OtOTPEgState *s)
 
+    if (partition >= OTP_PART_SECRET0 && partition <= OTP_PART_SECRET2) {
+        OtOTPTokenValue key = scrambling_keys[partition - OTP_PART_SECRET0];
+        uint64_t data = ((uint64_t)data_hi << 32) | data_lo;
```
nit pick: 32u

rivos-eblot commented on hw/opentitan/ot_otp_eg.c:
```diff
@@ -1762,6 +1768,17 @@ static void ot_otp_eg_dai_read(OtOTPEgState *s)
+        OtOTPTokenValue key = scrambling_keys[partition - OTP_PART_SECRET0];
+        uint64_t data = ((uint64_t)data_hi << 32) | data_lo;
+        OtPresentState *ps = ot_present_new();
+        ot_present_init(ps, (uint8_t *)&key);
```
nit pick: `(const uint8_t*)`

AlexJones0 committed:
[ot] hw/opentitan: ot_otp_eg: Load OTP scrambling keys

Load the OTP scrambling keys from hex-string configured properties.
These keys are used for decoding the scrambled data stored in secret
partitions (`SECRET0`, `SECRET1`, etc.)

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] hw/opentitan: ot_otp_dj: Load OTP scrambling keys

Load the OTP scrambling keys from hex-string configured properties.
These keys are used for decoding the scrambled data stored in secret
partitions (`SECRET0`, `SECRET1`, etc.)

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] hw/opentitan: ot_otp_dj: perform partition descrambling in DAI

Unscramble DAI reads from secret partitions, which are scrambled with a
scrambling key (configured per-partition as a netlist constant) via a
simple Present block cipher.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] scripts/opentitan: cfggen.py: Load OTP scrambling keys from SV

Parse the hard-coded OTP scrambling key constants out of the relevant SV
partition SV files into the generated OT configuration files, so that
they can be used in the OTP.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] docs/config: opentitan: Regenerate default config files

These files need to include the new OTP scrambling keys, so regenerate
them from the latest commits on the `master` branch (for Darjeeling) and
the `earlgrey_1.0.0` branch (for Earlgrey).

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 commented:
This should be working on Earlgrey going via `cfggen.py` to load the constants (at least, `//sw/device/tests:otp_ctrl_descrambling_test_sim_qemu_rom_with_fake_keys` passes when https://github.com/lowRISC/opentitan/pull/28541 is merged upstream).

I've tried to add the support for Darjeeling as well, but don't have an easy way to test it. I can see the keys are being parsed into the configuration file correctly at least. @rivos-eblot would it be possible for you could check the Darjeeling implementation?

rivos-eblot reviewed (approved):
LGTM

rivos-eblot commented:
> . @rivos-eblot would it be possible for you could check the Darjeeling implementation?

I can give it a try, but I do not know how. Let me check if we have some way to test it and get back to you.
Meanwhile, I think it is safe to merge this PR. Thanks.

AlexJones0 commented on python/qemu/ot/otp/const.py:
```diff
@@ -221,6 +228,22 @@ def get_digest_pair(self, name: str, prefix: str) -> dict[str, str]:
+        except KeyError as exc:
+            raise ValueError('No scrambling key constants found') from exc
+        if len(key_values) <= idx:
+            raise ValueError(f'No such key {name} in the key array') from exc
```
Ah, I tried to push this change but you had already merged the PR 😅
```suggestion
            raise ValueError(f'No such key {name} in the key array')
```
It'll likely just throw an `UnboundLocalError` instead of a `ValueError` on this error case, so not a huge deal.

I'll open a PR to fix this.
